### PR TITLE
Issue #14

### PR DESCRIPTION
**Pull request by fadlinurhasan**

### CODE REVIEW

File: commands.go [hunk 1]
```diff
@@ -73,6 +75,9 @@ func runAgent() {
 	provisionAgent := agent.NewProvisionAgent(hostname, daemon, pfclient)
 	go provisionAgent.Run()
 
+	bootstrapAgent := agent.NewBootstrapAgent(hostname, fullPath, daemon, pfclient)
```
Comment (giosakti) [inline commands.go]: If we do it like this, then provision and bootstrap might happen simultaneously (which is not what we want, race condition may occur)

I think the logic of bootstrap should be put inside provision logic itself,
it should happen after provision process is finished

File: commands.go [hunk 2]
```diff
@@ -34,6 +34,8 @@ func CmdAgent(ctx *cli.Context) {
 func runAgent() {
 	hostname, _ := os.Hostname()
 	ipaddress := getLocalIP()
+	filename := util.RandomString(10)
```
Comment (giosakti) [inline commands.go]: hmm yeah, I think we should generate this filename & fullpath within bootstrap agent. is that possible?

File: daemon/lxd.go [hunk 3]
```diff
@@ -166,3 +169,92 @@ func (l *LXD) DeleteContainer(hostname string) (bool, error) {
...
+	var content string
+	mode := 644
+	contentType := "file"
+	writeMode := "overwrite"
+
+	// // prepare filescript
+	for _, bs := range container.Bootstrappers {
+		switch bs.Type {
+		case "chef-solo":
+			content = `
```
Comment (giosakti) [inline daemon/lxd.go]: can content of the file be put as a file somewhere else and read on the fly here?

File: daemon/lxd.go [hunk 4]
```diff
@@ -166,3 +169,92 @@ func (l *LXD) DeleteContainer(hostname string) (bool, error) {
 
 	return true, nil
 }
+
+func (l *LXD) CreateContainerFile(container pfmodel.Container, fullPath string) error {
```
Comment (giosakti) [inline daemon/lxd.go]: please, explicit naming.. what is `CreateContainerFile`? it's not clear enough

File: daemon/lxd.go [hunk 5]
```diff
@@ -166,3 +169,92 @@ func (l *LXD) DeleteContainer(hostname string) (bool, error) {
...
+	var content string
+	mode := 644
+	contentType := "file"
+	writeMode := "overwrite"
+
+	// // prepare filescript
```
Comment (giosakti) [inline daemon/lxd.go]: unnecessary comment

File: daemon/lxd.go [hunk 6]
```diff
@@ -166,3 +169,92 @@ func (l *LXD) DeleteContainer(hostname string) (bool, error) {
...
+	err = l.targetSrv.CreateContainerFile(container.Hostname, fullPath, fileArgs)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (l *LXD) ExecContainer(container pfmodel.Container, fullPath string) (bool, error) {
+	//  pushing file to container
```
Comment (giosakti) [inline daemon/lxd.go]: unnecessary comment

File: daemon/lxd.go [hunk 7]
```diff
@@ -166,3 +169,92 @@ func (l *LXD) DeleteContainer(hostname string) (bool, error) {
...
+	}
+
+	// Setup the exec arguments (fds)
+	args := client.ContainerExecArgs{
+		Stdin:  os.Stdin,
+		Stdout: os.Stdout,
+		Stderr: os.Stderr,
+	}
+
+	// Get LXD to create the container (background operation)
```
Comment (giosakti) [inline daemon/lxd.go]: inappropriate comment?

File: agent/provision_agent.go [hunk 8]
```diff
@@ -42,7 +42,7 @@ func (a *provisionAgent) Run() {
 }
 
 func (a *provisionAgent) Process() bool {
-	scs, err := a.pfclient.FetchContainersFromServer(a.nodeHostname)
+	scs, err := a.pfclient.FetchContainersFromServer(a.nodeHostname, "ListScheduledContainers")
```
Comment (giosakti) [inline agent/provision_agent.go]: Hmm.. now I think we should remove if else logic in this method
it's bad to sent `ListScheduledContainers` as an argument

File: daemon/lxd.go [hunk 9]
```diff
@@ -166,3 +169,92 @@ func (l *LXD) DeleteContainer(hostname string) (bool, error) {
...
+
+	err = l.targetSrv.CreateContainerFile(container.Hostname, fullPath, fileArgs)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (l *LXD) ExecContainer(container pfmodel.Container, fullPath string) (bool, error) {
```
Comment (giosakti) [inline daemon/lxd.go]: please fix the naming, not showing intent

File: daemon/lxd_test.go [hunk 10]
```diff
@@ -1,11 +1,16 @@
 package daemon
 
 import (
+	"fmt"
+	"os"
 	"testing"
+	// "io/ioutil"
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply daemon/lxd_test.go]: Done

File: util/file.go [hunk 11]
```diff
@@ -0,0 +1,20 @@
+package util
+
+import (
+	"io"
+	"os"
+)
+
+func WriteToFile(filename string, data string) (*os.File, error) {
```
[the remark this answers is not in the document]
Comment (fadlinurhasan) [reply util/file.go]: done. since we create new file. so, we use `WriteStringToFile`

File: commands.go (hunk 1 above)
Comment (fadlinurhasan) [reply commands.go]: Done

File: commands.go (hunk 2 above)
Comment (fadlinurhasan) [reply commands.go]: Done

File: daemon/lxd.go (hunk 3 above)
Comment (fadlinurhasan) [reply daemon/lxd.go]: since we need bootstrapper data. now, we move it to be different func. please check

File: daemon/lxd.go (hunk 4 above)
Comment (fadlinurhasan) [reply daemon/lxd.go]: Done

File: daemon/lxd.go (hunk 5 above)
Comment (fadlinurhasan) [reply daemon/lxd.go]: Done

File: daemon/lxd.go (hunk 6 above)
Comment (fadlinurhasan) [reply daemon/lxd.go]: Done

File: daemon/lxd.go (hunk 7 above)
Comment (fadlinurhasan) [reply daemon/lxd.go]: done

File: agent/provision_agent.go (hunk 8 above)
Comment (fadlinurhasan) [reply agent/provision_agent.go]: Done.

File: daemon/lxd.go (hunk 9 above)
Comment (fadlinurhasan) [reply daemon/lxd.go]: done

File: util/file.go [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+	defer file.Close()
+
+	_, err = io.WriteString(file, data)
+	if err != nil {
+		return file, err
+	}
+	return file, nil
+}
+
+func SetupBootstrapFileContent(bs pfmodel.Bootstrapper, content string, mode int) (string, int) {
```
Comment (giosakti) [inline util/file.go]: I think GenerateBootstrapFileContent as a name would show more intent

File: daemon/lxd.go [hunk 13]
```diff
@@ -166,3 +169,75 @@ func (l *LXD) DeleteContainer(hostname string) (bool, error) {
...
 }
+
+func (l *LXD) CreateContainerBootstrapFile(container pfmodel.Container, fullPath string) error {
+	var content string
+	mode := 644
+	contentType := "file"
+	writeMode := "overwrite"
+
+	for _, bs := range container.Bootstrappers {
+		content, mode = util.SetupBootstrapFileContent(bs, content, mode)
```
Comment (giosakti) [inline daemon/lxd.go]: I don't understand the flow on this function,
does it means that if we have more than one bootstrappers, only the last bootstrappers will be written on the container?

Comment (giosakti) [reply daemon/lxd.go]: Also what is the purpose of setting mode to `644` above? it seems the `SetupBootstrapFile` method override that

File: util/file.go [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+	switch bs.Type {
+	case "chef-solo":
+		content = `
```
Comment (giosakti) [inline util/file.go]: can we put this content into an external template file instead and not hardcode it here?

File: daemon/lxd.go [hunk 15]
```diff
@@ -166,3 +169,75 @@ func (l *LXD) DeleteContainer(hostname string) (bool, error) {
...
+		Type:      contentType,
+		WriteMode: writeMode,
+	}
+
+	err = l.targetSrv.CreateContainerFile(container.Hostname, fullPath, fileArgs)
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (giosakti) [inline daemon/lxd.go]: I think we should remove the temporary file created

File: daemon/lxd.go [hunk 16]
```diff
@@ -166,3 +169,75 @@ func (l *LXD) DeleteContainer(hostname string) (bool, error) {
 
 	return true, nil
 }
+
+func (l *LXD) CreateContainerBootstrapFile(container pfmodel.Container, fullPath string) error {
```
Comment (giosakti) [inline daemon/lxd.go]: Instead of passing fullPath here, should we just generate the random path in this function instead?

File: daemon/lxd.go [hunk 17]
```diff
@@ -166,3 +169,75 @@ func (l *LXD) DeleteContainer(hostname string) (bool, error) {
...
+
+	err = l.targetSrv.CreateContainerFile(container.Hostname, fullPath, fileArgs)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (l *LXD) ExecContainerBootstrap(container pfmodel.Container, fullPath string) (bool, error) {
```
Comment (giosakti) [inline daemon/lxd.go]: can you rename `fullPath` it is not showing intent
this is the path of bootstrap file right?

File: agent/provision_agent.go [hunk 18]
```diff
@@ -172,3 +175,18 @@ func (a *provisionAgent) deleteContainer(sc pfmodel.Container, lcs *pfmodel.Cont
 
 	return true, nil
 }
+
+func (a *provisionAgent) bootstrapProcess() bool {
+	log.WithFields(log.Fields{}).Warn("Starting bootstrap agent...")
```
Comment (giosakti) [inline agent/provision_agent.go]: Reword this to `Bootstrapping...`
we don't actually start another agent because it runs in the same goroutine

File: agent/provision_agent.go [hunk 19]
```diff
@@ -172,3 +175,18 @@ func (a *provisionAgent) deleteContainer(sc pfmodel.Container, lcs *pfmodel.Cont
...
+	filename := util.RandomString(10)
```
Comment (giosakti) [inline agent/provision_agent.go]: I think we should remove this temporary filepath creation and moved it to within the function instead

File: agent/bootstrap_agent.go [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+
+	return &bootstrapAgent{
+		nodeHostname:    nodeHostname,
+		fullPath:        fullPath,
+		containerDaemon: containerDaemon,
+		pfclient:        pfclient,
+	}
+}
+
+func (a *bootstrapAgent) Run() {
```
Comment (giosakti) [inline agent/bootstrap_agent.go]: Is this still necessary?
